File: backend/models/purchase.py
# =============================================================================
# File: purchase.py
# Project: Grocery Intelligence Platform
# Author: Anita Woodford
# Description: Defines purchase models for tracking store purchase facts before report calculations.
# Security Note: This model stores purchase metadata only and must not store payment credentials.
# SRS Traceability: Supports SRS v5.0 purchase history, coupon tracking, reward tracking, and report input requirements.
# SDD Traceability: Supports SDD v5.0 backend model validation and data access layer purchase documents.
# =============================================================================

from __future__ import annotations  # Enables forward-compatible type annotations.

from datetime import date  # Imports the date type used for purchase dates.
from decimal import Decimal  # Imports Decimal for exact money values.
from enum import StrEnum  # Imports StrEnum so enum values behave like strings.
from typing import Optional  # Imports Optional for fields that are not required in Phase 1.

from pydantic import ConfigDict  # Imports ConfigDict to control Pydantic model behavior.
from pydantic import Field  # Imports Field to define defaults and validation constraints.

from backend.models.base import BaseDocument  # Imports the shared Mongo-style base document model.
from backend.models.base import PyObjectId  # Imports the shared Mongo ObjectId-compatible type.
from backend.models.rebate import RebateStatus  # Imports the existing rebate lifecycle enum.


class InputType(StrEnum):  # Defines the allowed sources for purchase input records.
    """Allowed source types for purchase records."""  # Documents the purpose of the enum.

    ONLINE_RECEIPT = "ONLINE_RECEIPT"  # Represents a purchase parsed from an online receipt.
    EMAIL_RECEIPT = "EMAIL_RECEIPT"  # Represents a purchase parsed from an email receipt.
    PDF = "PDF"  # Represents a purchase parsed from a PDF file.
    PHOTO = "PHOTO"  # Represents a purchase parsed from a receipt photo.
    CSV_IMPORT = "CSV_IMPORT"  # Represents a purchase imported from a CSV file.
    MANUAL_ENTRY = "MANUAL_ENTRY"  # Represents a purchase entered manually by the user.
    COPY_PASTE = "COPY_PASTE"  # Represents a purchase created from pasted receipt text.


class CouponSource(StrEnum):  # Defines the allowed coupon-source values.
    """Allowed coupon sources for purchase records."""  # Documents the purpose of the enum.

    WALGREENS_DIGITAL = "WALGREENS_DIGITAL"  # Represents a Walgreens digital coupon.
    CVS_DIGITAL = "CVS_DIGITAL"  # Represents a CVS digital coupon.
    HEB_DIGITAL = "HEB_DIGITAL"  # Represents an H-E-B digital coupon.
    WALMART_DIGITAL = "WALMART_DIGITAL"  # Represents a Walmart digital coupon.
    MANUFACTURER = "MANUFACTURER"  # Represents a manufacturer coupon.
    PAPER = "PAPER"  # Represents a paper coupon.


class RewardUsed(BaseDocument):  # Defines a reusable reward-used value object.
    """Tracks whether a store reward was used and how much value was applied."""  # Explains the model purpose.

    model_config = ConfigDict(extra="forbid")  # Rejects undefined fields during validation.

    used: bool = Field(default=False)  # Stores whether this reward type was used.
    amount: Decimal = Field(default=Decimal("0.00"), ge=Decimal("0.00"))  # Stores the reward amount used.


class PurchaseModel(BaseDocument):  # Defines the purchase document stored by the backend.
    """Stores purchase facts before later report-level calculations."""  # Explains the model purpose.

    model_config = ConfigDict(extra="forbid")  # Rejects undefined fields such as final_cost_after_rebates.

    product_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links this purchase to a product document.
    store_ref: PyObjectId = Field(...)  # Links this purchase to the store where it happened.
    canonical_name: str = Field(..., min_length=1)  # Stores the normalized product name.
    raw_name: Optional[str] = Field(default=None)  # Stores the original receipt text when available.
    category: str = Field(..., min_length=1)  # Stores the product category.
    brand: Optional[str] = Field(default=None)  # Stores the product brand when available.
    size: Optional[str] = Field(default=None)  # Stores the product size when available.
    quantity: int = Field(default=1, ge=1)  # Stores the purchased quantity.
    purchase_date: date = Field(...)  # Stores the date the purchase occurred.

    shelf_price: Decimal = Field(..., ge=Decimal("0.00"))  # Stores the regular shelf price before sales.
    sale_price: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Stores the sale price when available.
    register_price: Decimal = Field(..., ge=Decimal("0.00"))  # Stores the register item total before register coupons.
    subtotal_before_coupons: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Stores basket subtotal before coupons.
    out_of_pocket: Decimal = Field(..., ge=Decimal("0.00"))  # Stores actual register cost after register coupons.

    coupon_used: bool = Field(default=False)  # Stores whether a coupon was used.
    coupon_source: Optional[CouponSource] = Field(default=None)  # Stores the coupon source when a coupon was used.
    coupon_amount: Decimal = Field(default=Decimal("0.00"), ge=Decimal("0.00"))  # Stores the coupon amount applied.

    register_rewards_used: RewardUsed = Field(default_factory=RewardUsed)  # Stores Walgreens Register Reward usage.
    cvs_ecb_used: RewardUsed = Field(default_factory=RewardUsed)  # Stores CVS ExtraBucks usage.

    rebate_company: Optional[str] = Field(default=None)  # Stores the rebate company name when available.
    rebate_amount: Decimal = Field(default=Decimal("0.00"), ge=Decimal("0.00"))  # Stores expected or recorded rebate value.
    rebate_status: Optional[RebateStatus] = Field(default=None)  # Stores rebate lifecycle status when available.

    source_type: InputType = Field(...)  # Stores how the purchase record entered the system.
    raw_input_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links to the raw receipt input document.
    parse_confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)  # Stores parser confidence when available.
    user_corrected: bool = Field(default=False)  # Stores whether the user corrected the parsed data.
    notes: str = Field(default="")  # Stores user or system notes.

    walgreens_cash_redeemed: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Walgreens Cash redeemed.
    walgreens_cash_earned: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Walgreens Cash earned.
    walgreens_cash_earned_by_item_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links Walgreens Cash earned to an item.
    rr_used_amount: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Register Rewards used amount.
    rr_earned_amount: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Register Rewards earned amount.
    rr_earned_by_item_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links Register Rewards earned to an item.
    walmart_cash_earned: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Walmart Cash earned.
    walmart_cash_earned_by_item_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links Walmart Cash earned to an item.
    walmart_cash_used: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores Walmart Cash used.
    ecb_earned: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores CVS ExtraBucks earned.
    ecb_earned_by_item_ref: Optional[PyObjectId] = Field(default=None)  # Optionally links CVS ExtraBucks earned to an item.
    ecb_used: Optional[Decimal] = Field(default=None, ge=Decimal("0.00"))  # Optionally stores CVS ExtraBucks used.

File: frontend/src/types/purchase.ts
import type { BaseDocument, ObjectId, StandardResponse } from './base'; // Imports shared document, ObjectId, and API response types.

export type InputType = // Defines allowed purchase input source values.
  | 'ONLINE_RECEIPT' // Represents an online receipt input.
  | 'EMAIL_RECEIPT' // Represents an email receipt input.
  | 'PDF' // Represents a PDF input.
  | 'PHOTO' // Represents a photo input.
  | 'CSV_IMPORT' // Represents a CSV import input.
  | 'MANUAL_ENTRY' // Represents manual user entry.
  | 'COPY_PASTE'; // Represents pasted text input.

export type CouponSource = // Defines allowed coupon source values.
  | 'WALGREENS_DIGITAL' // Represents a Walgreens digital coupon.
  | 'CVS_DIGITAL' // Represents a CVS digital coupon.
  | 'HEB_DIGITAL' // Represents an H-E-B digital coupon.
  | 'WALMART_DIGITAL' // Represents a Walmart digital coupon.
  | 'MANUFACTURER' // Represents a manufacturer coupon.
  | 'PAPER'; // Represents a paper coupon.

export interface RewardUsed { // Defines a reward amount used at checkout.
  used: boolean; // Stores whether the reward was used.
  amount: string; // Stores the reward amount as a Decimal-safe string.
}

export interface PurchaseModel extends BaseDocument { // Defines one saved purchase record.
  product_ref?: ObjectId | null; // Stores the linked product id when available.
  store_ref: ObjectId | string; // Stores the linked store id or Phase 1 store key.
  canonical_name: string; // Stores the normalized item name.
  raw_name?: string | null; // Stores the original parsed item name when available.
  category: string; // Stores the item category.
  brand?: string | null; // Stores the item brand when available.
  size?: string | null; // Stores the item size when available.
  quantity: number; // Stores the purchased quantity.
  purchase_date: string; // Stores the purchase date as YYYY-MM-DD.
  shelf_price: string; // Stores the regular shelf price as a Decimal-safe string.
  sale_price?: string | null; // Stores the sale price as a Decimal-safe string when available.
  register_price: string; // Stores the register price before rebates as a Decimal-safe string.
  subtotal_before_coupons?: string | null; // Stores basket subtotal before coupons when available.
  out_of_pocket: string; // Stores actual checkout cost as a Decimal-safe string.
  coupon_used?: boolean; // Stores whether a coupon was used.
  coupon_source?: CouponSource | null; // Stores the coupon source when applicable.
  input_type?: InputType | null; // Stores how the purchase entered the system.
  raw_input_ref?: ObjectId | null; // Stores the raw input archive reference.
  parser_version?: string | null; // Stores parser version for traceability.
  parse_confidence?: number | null; // Stores parse confidence when available.
  user_corrected?: boolean; // Stores whether the user corrected this purchase.
  notes?: string | null; // Stores optional user notes.
}

export interface CreatePurchaseRequest { // Defines request body for creating a purchase.
  store_ref: ObjectId | string; // Sends the store id or Phase 1 store key.
  canonical_name: string; // Sends the normalized item name.
  category: string; // Sends the product category.
  quantity: number; // Sends the purchase quantity.
  purchase_date: string; // Sends the purchase date as YYYY-MM-DD.
  shelf_price: string; // Sends shelf price as a Decimal-safe string.
  register_price: string; // Sends register price as a Decimal-safe string.
  out_of_pocket: string; // Sends out-of-pocket amount as a Decimal-safe string.
  sale_price?: string | null; // Sends sale price when available.
  brand?: string | null; // Sends brand when available.
  size?: string | null; // Sends size when available.
  coupon_used?: boolean; // Sends whether a coupon was used.
  coupon_source?: CouponSource | null; // Sends coupon source when available.
  notes?: string | null; // Sends optional notes.
}

export interface UpdatePurchaseRequest { // Defines request body for editing a purchase.
  canonical_name?: string; // Updates normalized item name.
  category?: string; // Updates category.
  quantity?: number; // Updates quantity.
  purchase_date?: string; // Updates purchase date.
  shelf_price?: string; // Updates shelf price.
  sale_price?: string | null; // Updates sale price.
  register_price?: string; // Updates register price.
  out_of_pocket?: string; // Updates out-of-pocket amount.
  brand?: string | null; // Updates brand.
  size?: string | null; // Updates size.
  coupon_used?: boolean; // Updates coupon-used flag.
  coupon_source?: CouponSource | null; // Updates coupon source.
  notes?: string | null; // Updates notes.
}

export type PurchaseResponse = StandardResponse<PurchaseModel>; // Defines one-purchase API response.
export type PurchaseListResponse = StandardResponse<PurchaseModel[]>; // Defines purchase-list API response.